[c10d] Integrate ncclGather into PT

fduwjj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #164267
* __->__ #164270
* #164265
* #164276

NCCL 2.28.3 now supports `ncclGather`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

fduwjj committed:
[c10d] Integrate ncclGather into PT

[ghstack-poisoned]

fduwjj commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -1084,10 +1084,30 @@ void gather(
+        "root must provide inputs.size()==numranks");
+    // Allocate one flat buffer [world_size * count]
+    flat = at::empty({numranks * count}, inputs.options());
+    recv_ptr = flat.data_ptr();
```
I see. thanks!

kr4chinin replied:
@coderabbitai Оставляем как есть — осознанный defer. Точечный паттерн фона OG-карточки рендерится через инлайновый `<img>` с `alt=""` + `eslint-disable`, потому что satori (`next/og`) не поддерживает `mask-image`/CSS-фоны. Картинка чисто декоративная, осмысленный alt всей OG задаётся через route-export `alt`, а не внутренний `<img>`. Тред оставляем открытым как зафиксированное решение, без правок.

fduwjj committed:
Update on "[c10d] Integrate ncclGather into PT"


NCCL 2.28.3 now supports `ncclGather`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

[ghstack-poisoned]

Skylion007 commented on torch/csrc/cuda/nccl.cpp:
```diff
@@ -1084,10 +1084,30 @@ void gather(
+  void* recv_ptr = nullptr;
+  at::Tensor flat; // keep alive until after NCCL call
+  if (cur_rank == root) {
+    TORCH_CHECK(
+        (int)outputs.size() == numranks,
+        "root must provide inputs.size()==numranks");
```
```suggestion
    TORCH_CHECK_VALUE(
        static_cast<int>(outputs.size()) == numranks,
        "root must provide inputs.size()==numranks");

fduwjj committed:
Update on "[c10d] Integrate ncclGather into PT"


NCCL 2.28.3 now supports `ncclGather`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

[ghstack-poisoned]

fduwjj committed:
Update on "[c10d] Integrate ncclGather into PT"


NCCL 2.28.3 now supports `ncclGather`, this PR aims at integrating it. (release note: https://docs.nvidia.com/deeplearning/nccl/release-notes/rel_2-28-3.html#rel_2-28-3)

[ghstack-poisoned]

fduwjj commented:
synced offline, we plan to have a private API called `_gather_into_tensor` and keep the current gather intact.

Skylion007 commented:
> synced offline, we plan to have a private API called `_gather_into_tensor` and keep the current gather intact.

Out of curosity why? is the functionality not the same as the prev API?
